### PR TITLE
Job cancelling

### PR DESCRIPTION
**Pull request by mwhamgenomics**
ClusterExecutors now implement a `__del__` method, which stops any running cluster jobs when the object is destroyed. This should allow Analysis-Driver runs to be stopped without leaving jobs hanging.

Executors and script writers have also been refactored.

Closes #29.

### CODE REVIEW

**Comment by tcezard** [comment]
Could you look at this thread and see if this could be better ?
http://stackoverflow.com/questions/323972/is-there-any-way-to-kill-a-thread-in-python

**Comment by mwhamgenomics** [comment]
We could do something like:

``` python
running_executors = {}

def stop_running_executors():
    for name, e in running_executors.items():
        e.cancel_job()

class ClusterExecutor:
    # ...
    def start(self):
        # start stuff...
        running_executors[self.job_id] = self

    def join(self):
        # join stuff...
        del running_executors[self.job_id]

    def cancel_job(self):
        # cancel stuff...
```

This would be:
- 👍 more explicit
- 👍 independent of `__del__`, which can be funny with garbage collection
- 👎 manual - we would have to make sure we catch terminate signals and explicitly call `stop_running_executors` in Analysis-Driver

Edit: we can keep both - explicit (safe) and via `__del__` (automatic)

**Comment by mwhamgenomics** [comment]
We forgot that ClusterExecutor is not a subclass of Thread - parallelisation happens when the job script is submitted, but the Python component remains single-threaded in this case. The problem, then, is not as complicated as we first thought

**Comment by mwhamgenomics** [comment]
`__del__` job cancellation is failing at the moment. Current hypothesis:
- call stack: `__del__` -> `self.cancel_job` -> `self._get_stdout` -> ... -> `sys.meta_path`
- storing the executor in `cluster_executor.running_executors` creates an additional reference to it
- this delays its cleanup, so `sys.meta_path` is cleaned up first, breaking `__del__`
- if storage in `cluster_executor.running_executors` is removed, `__del__` _sometimes_ works

This means that `__del__` is not guaranteed to cancel the job, so it's doing more harm than good in this case. We've decided to go with explicit job cancellation only.

File: egcg_core/executor/cluster_executor.py [hunk 1]
```diff
@@ -150,3 +168,7 @@ def _job_exit_code(self):
 
         self.info('Got %s states from %s jobs: %s', len(states), len(reports), states)
         return exit_status
+
+    def cancel_job(self):
+        msg = self._get_stdout('scancel ' + self.job_id)
```
Comment (tcezard) [inline egcg_core/executor/cluster_executor.py:177]: Don't we need to test _job_finished to avoid sending an scancel query to finished jobs ?

**Pull request merged**